### ISSUE TITLE
Collision and Interpreter timing test cases

### ISSUE THREAD

**Issue by fmatthew5876** [EasyRPG/Player#1626]
There is at least 1 bug with event collisions and move routes

Set an event to `Event Touch`, layer=below, a message in the code

Create a move route for the player which walks over the event but doesn't stop on it.

In RPG_RT, the player walks over the event and nothing happens.
In Player. the player walks over the event and the message box appears as the player steps over the event.

Now put an obstacle on the other side of the event

In RPG_RT, the player walks over the event, gets blocked, and nothing happens.
In Player. the player walks over the event, gets block, and the message box repeats

I will use this issue to study collision / move route behavior

**Comment by fmatthew5876** [comment]
Move route test case dimensions
-----------------

Move routes

1. Stops on event
2. Walks past event
3. Walks past event but blocked
4. Walks past event but blocked and skippable
5. Walks past event, blocked, and then later a stop all movement command happens
6. Walks past event, blocked, and then a new move route doesn't change position

Event Layers

1. below
2. same
3. above

Trigger types

1. action
2. player touch
3. event touch

move frequency
1. Freq=8 (no stop)
2. Freq=7 (stopped)

**Comment by fmatthew5876** [comment]
Now some frame timings to figure out exactly when the collision happens:

Setup a map event with `layer=below`, `trigger=collision`
Set the code to:
```
OpenSaveMenu
```

Start the player to the left of the event.

Move the player and step on the event

| Save | remaining step |
| -- | -- |
| 1 | 192 |
| 2 | 128 |
| 3 | 64 |
| 4 | 0 |

Result: The event starts immediately as soon as the player x,y position changes and repeats over and over again. There is a 1 frame delay between each execution of the event. The first frame of movement, `remaining_step=224` but our save doesn't happen until the next frame when `remaining_step=192`.

Now change `trigger=touched`

| Save | remaining step |
| -- | -- |
| 1 | 0 |

Result: the event doesn't trigger until player completes movement.

Setup this common parallel event
```
SetMoveRoute: Player, Right
```

| Save | remaining step | stop count |
| -- | -- | -- |
| 1 | 0 | 0 |

**Comment by fmatthew5876** [comment]
Tests for collision trigger
------------------------

Setup a map event with `layer=below`, `trigger=collision`
Set the code to:
```
OpenSaveMenu
```

Start the player to the left of the event.

Move the player and step on the event

| Save | remaining step |
| -- | -- |
| 1 | 192 |
| 2 | 128 |
| 3 | 64 |
| 4 | 0 |

Result: The event starts immediately as soon as the player x,y position changes and repeats over and over again. There is a 1 frame delay between each execution of the event. The first frame of movement, `remaining_step=224` but our save doesn't happen until the next frame when `remaining_step=192`.

Setup this common parallel event
```
SetMoveRoute: Player, Right
```

| Save | remaining step | stop count |
| -- | -- | -- |
| 1 | 0 | 0 |

Result: Event waits for player move route to complete

**Comment by Ghabry** [comment]
Also interesting to test is if our Jump without Land and Land without Jump logic is correct and if this results in event triggering or not.

**Comment by fmatthew5876** [comment]
Test Cases 3: Player touch timing logic -- When do player touch events happen in the frame?
-----------------

1. Setup an event with `layer=below`, `trigger=touch`

Set the code:
```
SetMoveRoute: ThisEvent, Right
OpenSaveMenu
```

Move the player and step on the event.

| Save | remaining step | stop count |
| -- | -- | -- |
| 1 | 0 | 65535 |

Result: Event movement happens immediately

2. Setup the event code with

```
SetMoveRoute: This Event, Move Right
```

Setup a parallel event with:

```
SetMoveRoute: Player, Move Right
Wait 0.1s
Loop:
  OpenSaveMenu
```

| Save | player remaining step | player stop count | event remaining step | event stop count | event running | 
| -- | -- | -- | -- | -- | -- |
| 1 | 64 | 0 | 0 | 7 | 0 | 
| 2 | 32 | 0 | 0 | 8 | 0 | 
| 3 | 0 | 0 | 0 | 65535 | 0 | 
| 4 | 0 | 0 | 240 | 0 | 0 | 
| 5 | 0 | 0 | 224 | 0 | 0 |

**Comment by fmatthew5876** [comment]
Test Cases 4: Triggered by decision key logic
------------------

1. Create a player touch event with this code:
```
OpenSaveMenu
```

| Action | triggered by decision key |
| -- | -- |
| Walk up to the event and step towards the event and save | 0 |
| Walk up to the event and press the decision key | 1 |
| Save your game from the menu after doing the previous | 1 |

Result: `triggered_by_decision_key` is only set when event is triggered and never cleared.

**Comment by fmatthew5876** [comment]
Test Cases 5: Effects on waiting_execution when we change page
--------

TBD

**Comment by fmatthew5876** [comment]
Test Cases 6: Same layer events
-------------------------------------------

1. Create an event with the code and `layer = same`
```
OpenSaveMenu
```

Change the trigger type as defined below and then control the player as instructed

| Action | trigger | Effect | triggered by decision key |
| -- | -- | -- | -- |
| Talk to | action | Event runs | 1 |
| Talk to | touch | Event runs | 1 |
| Talk to | collision | Event runs | 1 |
| Step into | action | | |
| Step into | touch | Event runs | 0 |
| Step into | collision | Event runs | 0 |

**Comment by fmatthew5876** [comment]
Test Cases 7: When does change page happen?
-----------------------------------------

1. Create 2 map events with 2 pages

Page 1 has `move_speed=1`, Page 2 has `move_speed=6`

EV01 page 2 activated by switch 1, EV02 page 2 activated by switch 2.

Set EV01 page 1 to autostart with this code:

```
OpenSaveMenu
OpenSaveMenu
Enable switch 2
OpenSaveMenu
OpenSaveMenu
Disable switch 2
OpenSaveMenu
OpenSaveMenu
```

| Save | EV01 move speed | EV02 move speed |
| -- | -- | -- |
| 1 | 1 | 1 |
| 2 | 1 | 1 |
| 3 | 1 | 6 |
| 4 | 1 | 6 |
| 5 | 1 | 1 |
| 6 | 1 | 1 |

2. Set EV01 back to `trigger=action`

Set EV02 page 1 to autostart with this code:

```
OpenSaveMenu
OpenSaveMenu
Enable switch 1
OpenSaveMenu
OpenSaveMenu
Disable switch 1
OpenSaveMenu
OpenSaveMenu
```

| Save | EV01 move speed | EV02 move speed |
| -- | -- | -- |
| 1 | 1 | 1 |
| 2 | 1 | 1 |
| 3 | 6 | 1 |
| 4 | 6 | 1 |
| 5 | 1 | 1 |
| 6 | 1 | 1 |

**Comment by fmatthew5876** [comment]
Test Cases 8 - Effects of Through Flag
--------------------------------------------------

1. Create EV01 with `layer=same` and `trigger=collision`
```
OpenSaveMenu
```

Create EV01 at autostart
```
SetMoveRoute: Player, Through On
EraseEvent
```

Step onto EV01

Result: Nothing happens

2. Put the player start position to the left of EV01. Set EV01 `move_type=horizontal`


Result: EV01 walks through player, nothing happens

3. Put an impassable tile under the player

Result: EV01 can't step onto the same tile as the player, and the event triggers repeatedly as if player through was false. :exclamation:

**Comment by fmatthew5876** [comment]
Test cases 9 - Effects of counter tile
-----------------------------------------------

1. Setup an event with a counter in front. 
```
Message: Hello
```

Try the following actions and event triggers

| Action | trigger | Effect |
| -- | -- | -- |
| Talk to counter | action | Event triggers |
| Talk to counter | touch | Nothing |
| Talk to counter | collision | Nothing |
| Step into counter | action | Nothing |
| Step into counter | touch | Nothing |
| Step into counter | collision | Nothing |

2. Now change the `trigger=action` and put 2 counters in front of the event

Talk to the counter

Result: Event triggers

3. Now change the `trigger=action` and put 3 counters in front of the event

Talk to the counter

Result: Event triggers


4. Now change the `trigger=action` and put 4 counters in front of the event

Talk to the counter

Result: Nothing

**Comment by fmatthew5876** [comment]
Test Cases 10 - Events with no code
-----------------------

1. Create 2 autostart events

EV01 with this code:
```
OpenSaveMenu
```

EV02 with no code:

| Save | EV01 waiting execution | EV02 waiting execution |
| -- | -- | -- |
| 1 | 0 | 0 |

Result: Events with no code don't ever get their waiting flag set.

**Comment by fmatthew5876** [comment]
Test Cases 11 - Triggering when interpreter is active
----------------------------------------

1. Setup EV01 with `layer=below`, `trigger=touch` to the right of the player
```
SetMoveRoute: ThisEvent
```

Setup EV02 as parallel,
```
Wait 0.1s
SetMoveRoute: Player, Move Right
Wait 0.1s
Loop:
  OpenSaveMenu
```

| Save | player remaining step | Player stop count | EV01 stop count | EV01 waiting execution |
| -- | -- | -- | -- | -- |
| 1 | 32 | 0 | 13 | 0 |
| 2 | 0 | 0 | 65535 | 0 |

Result: Event triggers

2. Change EV02 to autostart

| Save | player remaining step | Player stop count | EV01 stop count | EV01 waiting execution |
| -- | -- | -- | -- | -- |
| 1 | 64 | 0 | 2 | 0 |
| 2 | 32 | 0 | 13 | 0 |
| 3 | 0 | 0 | 0 | 1 | 

Result: Event is waiting for execution

**Comment by fmatthew5876** [comment]
Test Cases 12 - First frame behaviors
-------------------------------------------------

As Cherry mentioned on the "first frame" we only process common events and events but not the player or vehicles.

This happens on:
* First frame of new game
* First frame teleporting into new map
* First frame after editing the menu (Thanks Cherry, didn't know this one!)

1. Create EV01 as parallel
```
Loop:
  OpenSaveMenu
```

Create EV02 as autostart
```
Wait 1.0s
SetMoveRoute: This Event
```

| Save | EV02 stop count | EV02 waiting execution |
| -- | -- | -- |
| 1 | 1 | 1 |
| 2 | 65535 | 0 |

Result: EV02 gets marked as waiting on the first frame and doesn't execute until 2nd frame

**Comment by fmatthew5876** [comment]
Test cases 13 - Exact timing of move trigger
----------------------------------

1. Setup EV01 with `trigger=touch`, `layer=below` and the code
```
Wait 0.1s
```

Setup EV02 as autostart with
```
Wait 0.1s
SetMoveRoute: Player, Right
Wait 0.1s
Wait 0.0s
SetMoveRoute: EV01, Up
Loop:
  OpenSaveMenu
```

Place the player to the left of EV01

| Save | Player remaining step | EV01 remaining step | EV01 stop count | EV01 waiting execution |
| -- | -- | -- | -- | -- |
| 1 | 32 | 0 | 65535 | 0 |
| 2 | 0 | 240 | 0 | 0 |

2. Change EV02 to
```
Wait 0.1s
SetMoveRoute: Player, Right
Wait 0.1s
Wait 0.0s
OpenSaveMenu
SetMoveRoute: EV01, Up
Loop:
  OpenSaveMenu
```


| Save | Player remaining step | EV01 remaining step | EV01 stop count | EV01 waiting execution |
| -- | -- | -- | -- | -- |
| 1 | 32 | 0 | 0 | 0 |
| 2 | 0 | 0 | 65535 | 1 |

**Comment by fmatthew5876** [comment]
Test Cases 14 - Multiple action events
-----------------

1. Setup EV01 `trigger=action`, `layer=below`, EV02 to the right of EV01 `trigger-action`, `layer=same`
Code for EV01 is `Message: A`, code for EV02 is `Message: B`

Step on EV01 and face EV02, press the decision key

Result: EV02 faces left. "A" and then "B" messages.

2. Move EV02 to the right and put a counter in between EV01 and EV02.


Step on EV01 and face the counter, press the decision key

Result: EV02 faces left. "A" and then "B" messages.

**Comment by fmatthew5876** [comment]
Test Cases 15 - Event change direction when triggered
-----------

TBD

**Comment by fmatthew5876** [comment]
Test Cases 16 - effects of an active move route when collision happens
------------------------

1. Setup EV01 to the right of the player. `layer=same`, `trigger=collision`, `move_type=horizontal`
```
OpenSaveMenu
```

Setup EV02 as autostart with this code:
```
SetMoveRoute: Player, Repeat, Wait
```

Result: Event triggers

2. Change EV01 to `layer=below`

Result: No trigger

3. Now setup the tiles horizontally like this:

| Tile | Event |
| -- | -- |
| Grass | Player |
| Grass | |
| Grass | |
| Grass | EV01 |
| Water | |

Change EV02 to this:
```
SetMoveRoute: Player, Speed Down (x4), Move Right
```

Result:

EV01 collides with player while player is mid move, and triggers immediately

Conclusion:
* collision on above/below don't trigger if move route is active
* collision on same doesn't care if move route is active

**Comment by fmatthew5876** [comment]
Test Cases 17 - map boundary
------------------------------------

1. Create a 20x15 map with horizontal looping.

Put EV01 `trigger=action`, `layer=same`, code=`Message:A` on the leftmost tile.

Talk to the event from the left.

Result: "A"

2. Put a counter on the right-most tile at the same y level as EV01

Talk to the counter:

Result: "A"

3. Move EV01 on tile to the right, and put a counter where EV01 was

Talk to the counter:

Result: "A"

4. Change the map to loop

Talk to the counter:

Result: Nothing

5. Remove the counters and move EV01 back to the left.

Move to the rightmost tile with same Y level as EV01 and talk.

Result: Nothing

**Comment by fmatthew5876** [comment]
Test Cases 18 - Does calling menus prevent events execution?
--------------------------

1. Setup EV01 as parallel with code:
```
Wait 0.0s
MENU
EraseEvent
```

Setup EV02 as autostart with code:
```
Tint Screen: 100, 100, 200
```

| MENU | Effect |
| -- | -- |
| Save | Tinted Before |
| Menu | Tinted Before |
| Load | Tinted Before |
| Battle | Screen is tinted in battle |

So scene calling doesn't prevent execution, but we have this logic in Player

**Comment by fmatthew5876** [comment]
Test Cases 19 - Trying to determine why we skip a frame in between invocations of events
--------------------

1. Setup EV01 as autostart
```
SetMoveRoute: Player, Turn Right 90 degrees
```

Setup EV02 as parallel:
```
Loop:
  OpenSaveMenu
```

| Save | Player stop count | Player direction | save.event.commands.size() | current command |
| -- | -- | -- | -- | -- |
| 1 | 0 | 2 | 1 | 0 |
| 2 | 65535 | 2 | 31 | 1 |
| 3 | 0 | 3 | 1 | 0 |
| 4 | 65535 | 3 | 31 | 1 |
| 5 | 0 | 0 | 1 | 0 |
| 6 | 65535 | 0 | 31 | 1 |
| 7 | 0 | 1 | 1 |

**Comment by fmatthew5876** [comment]
Test Cases 20 - When is collision checked?
-----------------------------------------------

1. Setup EV01 as collision, below with this code
```
Var1 + 1
```

EV02 as collision, below with this code
```
Var2 + 1
```

Setup common event:
```
Set EV01 to player
Set EV02 to player
Loop:
  OpenSaveMenu
```

| Save | Var1 | Var2 | EV01 waiting execution | EV02 waiting execution |
| -- | -- | -- | -- | -- |
| 1 | 0 | 0 | 1 | 1 |
| 2 | 1 | 0 | 0 | 1 |
| 3 | 1 | 0 | 0 | 1 |

2. Delete the events

Setup EV01 as collision, below with this code
```
Var1 + 1
```

EV02 as autostart with this code
```
Set EV01 to Player
Wait 0.0s
Erase Event
```

Setup common event:
```
Loop:
  OpenSaveMenu
```

| Save | Var1  | EV01 waiting execution | EV02 waiting execution | current command |
| -- | -- | -- | -- | -- |
| 1 | 0 | 0 | 1 | 0 |
| 2 | 0 | 0 | 0 | 1 |
| 3 | 0 | 0 | 1 | 2 |
| 4 | 0 | 0 | 1 | 2 |
| 5 | 0 | 0 | 1 | 3 |
| 6 | 0 | 0 | 1 | 0 |
| 7 | 1 | 0 | 1 | 0 |

**Comment by fmatthew5876** [comment]
@CherryDT

Is it correct that if the interpreter is running or *any* event is waiting for foreground execution, we will not increment `stop_count` for events which don't have a move route?

That seems to be the behavior, but then I found an edge case on the first frame.

Create EV01 parallel map event with
```
OpenSaveMenu
```
and EV02 autostart event with
```
Wait 0.1s
```

Now start a new game and save on the first frame. You'll see EV02 `waiting_execution=1` and `stop_count=1`!

On the first frame we update map events, so it makes sense that `waiting_execution` gets set. But since we don't run foreground events, it stays set. The thing that confuses me is the `stop_count` here. Since we set `waiting_execution`, shouldn't `stop_count` be 0?

It seems on this first frame, we always increment `stop_count` Is there some special case logic for this?

I hope all that made sense..

**Comment by CherryDT** [comment]
Hm I don't think `waiting_execution` has anything to do with it.

As far as I understand, it's more like this in the updating method of the event:

* If `!active`, abort
* If no active page, abort
* If `processed`, abort
* `processed` is set to `true`
* If `remaining_step == 0`, then do the main action (e.g. next movement - this influences `remaining_step` itself)
* Update flash
* One of the following:
  * If `remaining_step > 0`, then advance current movement (again influences `remaining_step` itself, also sets `stop_count` to 0 at the end)
  * If `stop_count == 0 || move_route_overwrite || ((SaveSystem.message_continue_events || !foregroundEventIsExecuting) && !pause)`*, increase `stop_count`
* Update `anim_count`

*: `foregroundEventIsExecuting` is defined as "global autostart worker has an event script loaded", i.e. `Save.events.commands.length > 0`**

**: For completeness: `Save.events` is the global autostart worker, while `SaveMapEvent.event_data` and `SaveCommonEvent.event_data` are the parallel process workers

**Comment by fmatthew5876** [comment]
What is `pause` in your example?

**Comment by CherryDT** [comment]
`SaveMapEventBase.pause`

**Comment by fmatthew5876** [comment]
Test Cases 21 - Study of stop count behavior when events running
---------------------------------

1. Setup EV02 parallel as
```
Loop:
  OpenSaveMenu
```

Setup EV01 autostart as
```
Wait 1.0s
```

| Save | player stop count | EV01 stop count | EV02 stop count | EV01 waiting execution |
| -- | -- | -- | -- | -- |
| 1 | 0 | 1 | 1 | 1 |
| 2 | 1 | 1 | 2 | 0 |
| 3 | 1 | 1 | 2 | 1 |

2. Change EV02 to
```
OpenSaveMenu
Enable Sw1
Loop:
  OpenSaveMenu
```

Change EV01 to be activated by Sw1

| Save | player stop count | EV01 stop count | EV02 stop count | EV01 waiting execution | current command |
| -- | -- | -- | -- | -- | -- |
| 1 | 0 | 0 | 1 | 0 | 0 |
| 2 | 1 | 0 | 2 | 0 | 0 |
| 3 | 2 | 1 | 3 | 0 | 1 |
| 4 | 2 | 1 | 3 | 0 | 1 |

**Comment by fmatthew5876** [comment]
@CherryDT 

When exactly does event page change logic happen?

If you look at test case (2) in my previous post, on frame 2 when activate switch 1. This happens in a parallel event, so the switch gets activated before map events update during the frame.


Essentially I think it should do this on frame 2:
1. Parallel worker enabled SW1
2. EV01 page 1 is now active, stop count animates, autostart sets waiting flag
3. EV01 starts executing (current_command goes to 1)

Instead it looks like there is a delay of 1 frame inserted here? When exactly does page change happen and how does this insert frame delays?

**Comment by CherryDT** [comment]
* Main update method of EV01 runs, does nothing
* Main update method of EV02 runs
* Parallel worker of EV02 (which is called from the event's main update method) sets SW1 and reruns all map events' checks for their active page
* Foreground worker runs, doesn't find anything to do

Next frame:

* Main update method of EV01 runs, activates the autostart (sets `waiting_execution`)
* Main update method of EV02 runs, changes scene to save menu
* ...
* Foreground worker runs, loads EV02's script who had `waiting_execution` set, handles "Wait 1.0s" command in EV02

My explanation about the workers was a simplification, in general it is like this:

* `processed` flag gets reset for all map events
* Main update methods for common events run (they just check if they are parallel process and the condition is met and then **invoke the parallel process worker for them**)
* Main update methods for map events run (they do a bunch of things, described in my previous post - but actually even more, because after the check for whether there is any active event page, they will **run their parallel process worker** and then check once more for `active`)
* Message window updates
* Timer updates
* Screen effects update
* Pictures update
* **Foreground worker runs**
* Panorama updates

Note that there is one additional special case: The first three items of this list (`processed` flag getting reset, main update methods for common events and then map events are run) are also executed when a teleport has just been executed (right after the position of the party was updated) or from inside the menu scene when a teleport/escape was executed, right _before_ the screen fades in.

**Comment by fmatthew5876** [comment]
Ok so thats the difference, I'm running all parallel events interpreter's separately *before* running the main loops. What you said, it sounds like the parallel events run during the update loop of events.

So that means map event update procedure should look like this?

1. If !active or no page: abort
2. If processed: abort
3. Set processed=true
4. Run parallel event code
5. If !active or no page: abort
6. Remaining stuff you outlined...

And that would mean also, that if movement collision makes EV05's update routine run before EV04, then EV05's parallel worker would also run before EV04's that frame?

**Comment by CherryDT** [comment]
> And that would mean also, that if movement collision makes EV05's update routine run before EV04, then EV05's parallel worker would also run before EV04's that frame?
Hm, I believe that's the case, yes.

> 5. If !active or no page: abort
Actually I think, there it is _only_ if !active

One interesting thing that now came to my mind is that the action method is only called when `remaining_step == 0`, and it's that one which triggers foreground event execution - _including autostart_, so actually that would mean that if an event is moving around and gets an autostart event page enabled in the middle of the movement, it will only start _executing_ it once the step is finished...

**Comment by fmatthew5876** [comment]
> One interesting thing that now came to my mind is that the action method is only called when `remaining_step == 0`, and it's that one which triggers foreground event execution - _including autostart_, so actually that would mean that if an event is moving around and gets an autostart event page enabled in the middle of the movement, it will only start _executing_ it once the step is finished...

Wow thats super helpful, I'll make a test around this to confirm.

How about `trigger=collision` checks? Are there any restrictions on when that is done?

It looks like when `layer=same`, we check every time the event tries to move onto the same tile as the player and the movement fails.

For `layer=below` or `above`, I think we check every frame?

**Comment by fmatthew5876** [comment]
> Actually I think, there it is _only_ if !active

Wierd, so that means if the parallel event disables all pages, we'll still process a frame of movement?

**Comment by CherryDT** [comment]
So for the triggers... that would take me a long time to figure out :( because it seems like there are checks sprinkled all over the place, in various locations that are in the middle of processing something related to movement, with slightly different behavior (sometimes events checking towards the party, sometimes the other way round, sometimes only for one of the two touch triggers, sometimes for both).

What they have in common is that an active foreground worker will suppress all checks.

Furthermore, being in an airship will suppress some checks, and having `move_route_overwrite` enabled on the party will suppress some other checks.

¯\_(ツ)_/¯

> Wierd, so that means if the parallel event disables all pages, we'll still process a frame of movement?

Hm, possibly, but given that the event then usually has no move route (other than one given explicitely through event code), it shouldn't matter that much, I guess

**Comment by fmatthew5876** [comment]
Test Cases 22 - Autostart checks during active movement
---------------------

1. Create EV01

Page 1 `triggger=action`, `move_speed=6`
Page 2 `trigger=autostart`, `move_speed=6`
```
Wait 1.0s
```

EV02 parallel
```
SetMoveRoute: EV01, left
SW1 On
Loop:
  OpenSaveMenu
```

| Save | EV01 remaining step | EV01 stop count | EV01 waiting execution | current command |
| -- | -- | -- | -- | -- |
| 1 | 0 | 65535 | 0 | 0 |
| 2 | 128 | 0 | 0 | 1 |
| 3 | 0 | 0 | 0 | 1 |

Results:
Frame 1
EV01 runs - nothing
EV02 runs - activated SW1

Frame 2
EV01 runs - `remaining_step == 0` so activates autostart condition
EV02 runs

2. Change EV02 to

```
SetMoveRoute: EV01, left
OpenSaveMenu
SW1 On
Loop:
  OpenSaveMenu
```

| Save | EV01 remaining step | EV01 stop count | EV01 waiting execution | current command |
| -- | -- | -- | -- | -- |
| 1 | 0 | 65535 | 0 | 0 |
| 2 | 128 | 0 | 0 | 0 |
| 3 | 0 | 0 | 0 | 0 |
| 4 | 0 | 0 | 0 | 1 |

Result:
Frame 2:
EV01 runs - Performs first half ot step `remaining_step 256 -> 128`
EV02 runs - Sets SW1 On

Frame 3:
EV01 runs - Finishes step `remaining_step 128 -> 0`
EV02 runs

Frame 4:
EV01 runs - `remaining_step == 0` so activates autostart condition

**Comment by fmatthew5876** [comment]
Test Cases 23 - Movement failed and skip - vs same layer triggering
-------------------------

1. Setup EV01 with `trigger=collision`, `layer=same`, `direction=right`
```
OpenSaveMenu
```

Setup EV02 as parallel
```
SetMoveRoute: Player, Right
EraseEvent
```

Place hero to the left of EV01

Result: Player Stuck

2. Change EV01 `trigger=touch`

Result: Player Stuck

3. Change EV02 code to
```
SetMoveRoute: Player, Skip, Right
EraseEvent
```

Result: Nothing

4. Change EV01 `trigger=collision`

Result: Nothing

Conclusion
---------------

Player move route never triggers same layer events.

**Comment by fmatthew5876** [comment]
@CherryDT 

In test case 20.6 I saw some really strange.

On the first frame of the game, only map events update. But on this particular case, I saw on frame 1 the players `stop_count=1` and `processed=1`

I setup EV01 to put a move route on EV02. EV02 was collision and below, and stepped onto the player on frame 1.

What must have happened is that the movement collision triggered, namely that EV02 stepping onto player's tile caused player to be updated that frame. I don't know how else the player could update on frame 1.

How does this movement collision stuff work? I was under the assumption it only happens for a potentially failed move, but it looks like maybe it happens anytime event A wants to step on a tile that contains other events, regardless of whether they would actually block movement or not?

I'm super confused by this now..

What conditions cause movement collision logic to take effect and cause out of order event updates?

**Comment by fmatthew5876** [comment]
Test Cases 24 - make way
--------------------------------

1. Create paralel event EV01 `layer=below` `trigger=parallel` to the right of the player
```
SetMoveRoute: ThisEvent, Move left
OpenSaveMenu
```

| Save | player stop count |
| 1 | 1 |

Result: Event causes player to update first frame

**Comment by CherryDT** [comment]
For 20.6 that sounds right, since even though that special case code does not _clear_ the player's processed flag, it can still be triggered to update due to another event's collision if the flag was cleared to begin with, which would be the case in the first frame.

However what's interesting to me is why the player doesn't update regularly on the first frame, as I don't see any reason for this behavior right now, but maybe I'm just overlooking something

**Comment by fmatthew5876** [comment]
> However what's interesting to me is why the player doesn't update regularly on the first frame, as I don't see any reason for this behavior right now, but maybe I'm just overlooking something

Maybe we're confusing terminology, since 2 updates happen on the "first frame".

 What I specifically mean is at the very beginning when that special case logic updates only map events before transition in. During this time, the player normally does not update. 

After that, during the "normal" update cycle the player and everything else does update.

Frame 1:
1. Update only map events
2. Transition In
3. Update map events again
4. Update player
5. Update everything else

Frame 2:
* Update everything

My test case showed the collision logic causing an extra player update to happen during (1).

What is confusing to me is why does the movement collision logic triggers here. If event A is `layer=below`, it can't possibly get its movement blocked by the player. Yet despite this, the player is forced to update (so it can possibly move out of the way) when A wants to occupy it's tile.

*It would be really helpful if you could outline the cases that cause this "collision logic" to trigger other events update routines to run.* I've just identified for example that the `through` flag prevents it.

This logic will be hard to pin down with tests as its not easy to see what events run in which order with my testing method.

**Comment by CherryDT** [comment]
Whenever any kind of event or the player even _check_ whether a move is possible (regardless of what happens afterwards), the following logic is applied:

* If new position is outside of the map: Return false
* If this event is in "through" mode: Return true
* Unless jumping:
  * If this event's graphic is a tile and this event's priority is below hero, check if the event's movement would conflict _itself_ (because for example the upper-layer tile would not allow the event "on it" to move away in the direction the event tries to - however, it has to be taken into account that _the tile would move itself_), and if yes, set self-conflict flag
  * Check if the event can move away from the current tile in the given direction based on chipset settings (not counting self-conflict). If not: Return false
* Enumerate map events by ID.
  * If an event is found whose coordinates match the target of the movement, **trigger update method** (note that this happens just if there is an event, regardless of it would collide yet!). Then, check again, and if it's still at the same position, check if there would be a collision (with the full rules involving priority, forbid-overlap flag, etc.). There is a special case that involves the self-conflict flag, see below.
  * If the event was found to be colliding: Return false (note that this means that not all events which are on this tile may have their update triggered!)
* Check if the chipset allows moving _onto_ the new tile. If not: Return false (Note: During jumping, this will check only whether the tile is passable in general, not from any particular direction!)
* Otherwise: Return true

**Collision check:**

* If either this or the other event is `through`: Return false
* If either this or the other event is `flying`: Return false
* If either this or the other event is `!active`: Return false
* If both this and the other event are actual events (and not the party or vehicles) and either this or the other event is `overlap_forbidden`: Return true
* If the other event's priority is "same level as hero" **and the self-conflict flag is set** (see above): Return true
* If this and the other event's priorities are equal: Return true
* Otherwise: Return false

**Comment by fmatthew5876** [comment]
@CherryDT 

What conditions prevent player input (directional input for movement, decision key, ESC menu) from being checked?

We have this `IsAnyEventStarting()` method, which checks all common events and map events to see if any are waiting for foreground execution. I'm inclined to believe this is bogus and the entire concept can be removed, but I wanted to confirm with you first.

**Comment by CherryDT** [comment]
Conditions that prevent party movement on map

* Airship is currently lifting off or landing
* `Save.events.commands.length > 0` (Note that this condition will, if fulfilled, clear `menu_calling` before aborting)
* Message window is visible
* Party has `move_route_overwrite`
* Party has `pause` (irrelevant)
* `encounter_calling` flag was set and was processed now
* `menu_calling` flag was set and was processed now
* Any common or map event has `waiting_execution`

Conditions that prevent decision key on map

* All of the above
* `remaining_step > 0` (checked *after* potentially initiating movement from directional keys)

Conditions that prevent escape key on map

(This happens after all of the above, without any of the above conditions though)

* Menu is not allowed (of course)
* Message window is visible
* `Save.events.commands.length > 0`

(Note that this won't immediately invoke the menu, but instead set the `menu_calling` flag.)

**Comment by fmatthew5876** [comment]
And how about `menu_calling` and `encounter_calling`?

Why were these flags invented instead of just changing the scene directly like the event commands do?

When the player pressed ESC, it sets `menu_calling`. When does `menu_calling` get checked and the menu scene called? At the end of the same frame we detected ESC key?

Are there any special cases with `menu_calling` outside of that?

It looks like `encounter_calling` is only used for random encounters. Is there some frame delay on when `encounter_calling` is set and when the scene is changed to battle? Special caveats with this flag?

**Comment by CherryDT** [comment]
> Why were these flags invented instead of just changing the scene directly like the event commands do?

I _guess_ in order to give those things a lower priority than others, especially event-controlled ones, and maybe to allow events to react on ESC press too, before the menu is actually called...? I'm not sure.

> When does menu_calling get checked and the menu scene called? At the end of the same frame we detected ESC key?

At the next frame, at the point where my post said "menu_calling flag was set and was processed now" above.

> Are there any special cases with menu_calling outside of that? Anything which will later clear the flag? How does it interact if a scene call from an event happened on the same frame?

Yes, only for random encounters. It is cleared when the player position changes by something other than movement (usually teleport), when a foreground event is active, when the `encount_calling` was set and is being processed.

> It looks like encounter_calling is only used for random encounters. Is there some frame delay on when encounter_calling is set and when the scene is changed to battle? Special caveats with this flag?

It is processed at the point I wrote "encounter_calling flag was set and was processed now" in the post above. (So it has the same delay as `menu_calling`.) It is cleared when a foreground event is triggered.

**Comment by fmatthew5876** [comment]
How about diagonal movement? 

Does it do all the checks you mentioned only against the diagonal tile the event wants to move to? Or does it also do some for the vertical/horizontal adjacent tiles as well?

**Comment by fmatthew5876** [comment]
Also for vehicles.

What if I press Enter (or have an event command) to disembark the ship? Will it just run through the same algorithm?

What if I want to land the airship? Is it the same there? If so, exactly which frame does the calculation start?

**Comment by CherryDT** [comment]
For diagonal, it just checks the diagonal tile. And regarding the directional passability, that is then just like a jump - none of the directional passability settings apply, only the general passability.

I believe the vehicle "Enter" happens at the same time. I'm not sure though.

**Comment by fmatthew5876** [comment]
> For diagonal, it just checks the diagonal tile. And regarding the directional passability, that is then just like a jump - none of the directional passability settings apply, only the general passability.

Hmm, that doesn't seem right.

If I put a tile that blocks only Left and Up, and then try to step onto it from the upper left. I get blocked.
If I make the tile block only up, it lets me step onto the tile
If I make the tile block only left, it lets me step onto the tile

If I put a blocking tile to the right of the player. I can step lower right.
If I put a blocking tile below the player. I can step lower right.
If I put blocking tiles both to the right and below the player. I cannot step lower right.

**Comment by CherryDT** [comment]
Mmh seems like I was wrong there indeed. What you say makes a lot more sense anyway

**Comment by fmatthew5876** [comment]
Test Cases 25 - Events off the edge of the map
--------------------------------------------

1. Setup EV01 as `layer=same` `trigger=action` `Message: Hello`

2. Setup EV02 as autostart
```
Var1 = -1
Var2 = 8
SetEventLocation(EV01, Var1, Var2)
EraseEvent
```

Walk over to the edge of the map and talk to EV01

Result: "Hello"

2. Change EV01 `trigger=touch`

Walk over to the edge of the map and try to step off the edge onto EV01

Result: Hello

3. Change EV01 `trigger=collision`

Walk over to the edge of the map and try to step off the edge onto EV01

Result: Hello

Conclusion:
----------------

layer same trigger checks don't care about the edge of the map

**Comment by fmatthew5876** [comment]
Test Cases 26 - Collision Revisted
------------------------

This is a set of test cases to nail down exactly when collision below checks happen. There is one collision check done at the same time as touch, which we have covered and are sure about.

There is another set of collision checks that happens only when the foreground interpreter is not running and the player doesn't have a move route. I am not yet sure exactly when this check happens. It could be any combination of the below:

* Player before movement
* Player after movement
* Event before movement
* Event after movement

Here we will collect test cases (some of them repeats from before) to try to figure this out.

1. Setup EV01 with `layer=below`, `trigger=collision`
Set the code to:
```
OpenSaveMenu
```

Start the player to the left of the event.

1. Move the player and step on the event

| Save | remaining step | stop count |
| -- | -- | -- |
| 1 | 192 | 0 |
| 2 | 128 | 0 |
| 3 | 64 | 0 |
| 4 | 0 | 0 |
| 5 | 0 | 2 |
| 6 | 0 | 3 |

On the 2nd frame we trigger the event.

Frame breakdown
* EV01 update: (nothing)
* Player update: move 224 - collides with EV01
* EV01 update: (maybe trigger?)
* Player update: move 192 (maybe trigger?)

**Comment by fmatthew5876** [comment]
Test Cases 27 - Collision Revisited part 2 - step out of the way
---------------------

1. Setup EV01 with `layer=below`, `trigger=collision`
Set the code to:
```
OpenSaveMenu
```

Setup EV02 parallel with
```
Wait 0.0s
SetMoveRoute: Player, Right
Wait 0.0s
StopAllMovement
SetMoveRoute: EV02, Up
EraseEvent
```

Result: No Trigger


| Frame | Event | Action | position | remaining step | player move route overwritten |
| -- | -- | -- | -- | -- | -- |
| 1P | EV01 | wait | != | 0 | 0 |
| 1P | EV02 | | != | 0 | 0 |
| 1 | EV01 | | != | 0 | 0 |
| 1 | EV02 | set move route on player, wait | != | 0 | 1 |
| 1 | Player | move right | == | 224 | 1 |
| 2 | EV01 | | == | 0 | 1 |
| 2 | EV02 | stop all movement, set move route on EV02, erase | == | 0 | 0 |
| 2 | Player | move right | == | 192 | 0 |
| 3 | EV01 | move up | != | 240 | 0 |
| 3 | Player | move right | != | 160 | 0 |

2. Swap EV01 and EV02

Result: No Trigger

| Frame | Event | Action | position | remaining step | player move route overwritten |
| -- | -- | -- | -- | -- | -- |
| 1P | EV01 | wait | != | 0 | 0 |
| 1P | EV02 | | != | 0 | 0 |
| 1 | EV01 | set move route on player, wait | != | 0 | 1 |
| 1 | EV02 | | != | 0 | 1 |
| 1 | Player | move right  | == | 224 | 1 |
| 2 | EV01 | stop all movement, set move route on EV02, wait | == | 0 | 0 |
| 2 | EV02 | move up | != | 240 | 0 |
| 2 | Player | move right | != | 192 | 0 |

**Comment by fmatthew5876** [comment]
Test Cases 27 - Collision Revisited part 3 - set event location
---------------------

1. Setup EV01 with `layer=below`, `trigger=collision`
Set the code to:
```
OpenSaveMenu
```

Setup EV02 parallel with
```
Set EV01 to Player
EraseEvent
```

| Save | player stop count |
| -- | -- |
| 1 | 1 |

Result: Triggers on frame 1 after pre-update

| Frame | Event | Action | position | remaining step | player move route overwritten |
| -- | -- | -- | -- | -- | -- |
| 1P | EV01 |  | == | 0 | 0 |
| 1P | EV02 | sets ev01 to player, erase | == | 0 | 0 |
| 1 | EV01 | :question: | == | 0 | 0 |
| 2 | Player | :question: | == | 0 | 0 |

2. Swap EV01 and EV02

| Save | player stop count |
| -- | -- |
| 1 | 1 |

Result: Triggers on frame 1 after pre-update

| Frame | Event | Action | position | remaining step | player move route overwritten |
| -- | -- | -- | -- | -- | -- |
| 1P | EV01 | sets ev02 to player, erase | == | 0 | 0 |
| 1P | EV02 | | == | 0 | 0 |
| 1 | EV02 | :question: | == | 0 | 0 |
| 2 | Player | :question: | == | 0 | 0 |

**Comment by fmatthew5876** [comment]
Test Cases 27 - Collision Revisited part 4 - Step Through Tests
-----------------------

1. Setup EV01 as parallel
```
SetMoveRoute: Player, Wait
SetMoveRoute: EV02, Left, Left
WAIT
StopAllMovement
OpenSaveMenu
EraseEvent
```

Setup EV02 as collision, below with code
```
OpenSaveMenu
```

Now try different frame counts for WAIT

| WAIT | player stop_count | EV02 stop count | EV02 remaining step | trigger |
| -- | -- | -- | -- | -- |
| 1 | 2 | 0 | 224 | Y |
| 2 | 3 | 0 | 208 | Y |
| 14 | 15 | 0 | 16 | Y |
| 15 | 16 | 0 | 0 | Y |
| 16 | 17 | 1 | 0 :question: | Y |
| 17 | 18 | 0 | 224 | N |

Conclusion:
---------------

In this case, the event is not allowed to step off the player. When we insert a wait of 16 frames, we would expect the event to step off with `remaining_step=240`, but instead the collision triggers before the event can move, blocks the movement, and triggers.

**Comment by fmatthew5876** [comment]
Test Cases 28 - foreground events and boarding
-------------------

1. Setup EV01 with 2 pages
Page1 - nothing
Page2 - autostart, activated by SW1 `Wait 1.0s`

Setup EV02 as parallel
```
GetOnOffVehicle
Wait 0.0s
Wait 0.0s
OpenSaveMenu
Enable SW1
Loop:
  OpenSaveMenu
```

| Save | remaining step | boarding | aboard | EV01 waiting execution | current command |
| -- | -- | -- | -- | -- | -- |
| 1 | 192 | 1 | 0 | 0 | 0 |
| 2 | 160 | 1 | 0 | 0 | 0 |
| 3 | 128 | 1 | 0 | 0 | 0 |
| 4 | 96 | 1 | 0 | 1 | 0 |
| 5 | 64 | 1 | 0 | 1 | 0 |
| 6 | 32 | 1 | 0 | 1 | 0 |
| 7 | 0 | 0 | 1 | 1 | 1 |

Result: Foreground interpreter waits for boarding to finish before running

2. Delete the events

Setup EV01 as autostart with
```
GetOnOffVehicle
OpenSaveMenu
```

| Save | remaining step | boarding | aboard | EV01 waiting execution | current command |
| -- | -- | -- | -- | -- | -- |
| 1 | 0 | 0 | 1 | 1 | 2 |

Result: Currently executing foreground event also waits for boarding

**Comment by fmatthew5876** [comment]
Test Cases 28 - foreground events and ascending
-------------------

1. Setup EV01 with 2 pages
Page1 - nothing
Page2 - autostart, activated by SW1 `Wait 1.0s`

Setup EV02 as parallel
```
Set airship to player
GetOnOffVehicle
Wait 0.0s
Wait 0.0s
OpenSaveMenu
Enable SW1
Loop:
  OpenSaveMenu
```

| Save | remaining_ascent | aboard | EV01 waiting execution | current command |
| -- | -- | -- | -- | -- |
| 1 | 240 | 1 | 0 | 0 |
| 3 | 224 | 1 | 0 | 0 |
| 4 | 216 | 1 | 1 | 0 |
| 30 | 8 | 1 | 1 | 0 |
| 31 | 0 | 1 | 1 | 1 |

Result: Foreground interpreter waits for boarding to finish before running

2. Delete the events

Setup EV01 as autostart with
```
GetOnOffVehicle
OpenSaveMenu
```

| Save | remaining ascent | remaining descent | aboard | EV01 waiting execution | current command |
| -- | -- | -- | -- | -- |
| 1 | 0 | 0 | 1 | 1 | 3 |
| 2 | 0 | 0 | 0 | 1 | 5 |

Result: Currently executing foreground event also waits for ascent and descent

**Comment by fmatthew5876** [comment]
Reopened until the above behaviors are covered by unit tests
